File: app/src/main/java/de/metzgore/beansplan/util/di/modules/ScheduleDaoModule.java
package de.metzgore.beansplan.util.di.modules;

import android.content.Context;
import dagger.Module;
import dagger.Provides;
import de.metzgore.beansplan.dailyschedule.DailyScheduleDaoImpl;
import de.metzgore.beansplan.shared.DailyScheduleDao;
import de.metzgore.beansplan.shared.WeeklyScheduleDao;
import de.metzgore.beansplan.weeklyschedule.WeeklyScheduleDaoImpl;

import javax.inject.Singleton;

@Module(includes = ContextModule.class)
public class ScheduleDaoModule {

    private boolean cacheOnly;

    public ScheduleDaoModule(boolean cacheOnly) {
        this.cacheOnly = cacheOnly;
    }

    @Provides
    @Singleton
    public DailyScheduleDao providesDailyScheduleDao(Context context) {
        return new DailyScheduleDaoImpl(context, cacheOnly);
    }

    @Provides
    @Singleton
    public WeeklyScheduleDao providesWeeklyScheduleDao(Context context) {
        return new WeeklyScheduleDaoImpl(context, cacheOnly);
    }
}

File: app/src/main/java/de/metzgore/beansplan/util/di/modules/ScheduleDaoModule.kt
package de.metzgore.beansplan.util.di.modules

import androidx.room.Room
import android.content.Context
import dagger.Module
import dagger.Provides
import de.metzgore.beansplan.data.room.BeansPlanDb
import de.metzgore.beansplan.data.room.ScheduleRoomDao
import javax.inject.Singleton

@Module(includes = [(ContextModule::class)])
class ScheduleDaoModule(private val inMemory: Boolean) {

    @Singleton
    @Provides
    fun provideDb(context: Context) = if (inMemory) {
        Room.inMemoryDatabaseBuilder(context, BeansPlanDb::class.java).build()
    } else {
        Room.databaseBuilder(context, BeansPlanDb::class.java, "beansplan.db").fallbackToDestructiveMigration().build()
    }

    @Provides
    @Singleton
    fun provideScheduleRoomDao(db: BeansPlanDb): ScheduleRoomDao {
        return db.scheduleDao()
    }
}
